File: internal/leanoff/module.go
package leanoff

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// Module is one .lean file: its import name, its path, and (after
// ParseImports) its project-internal dependencies.
type Module struct {
	Name string
	Path string
	Deps []string
}

var importRe = regexp.MustCompile(`^\s*import\s+([A-Za-z_][A-Za-z0-9_.]*)`)

var skipDirs = map[string]bool{
	".lake": true, ".git": true, ".vscode": true, "node_modules": true,
	"dist": true, "out": true, "build": true, ".leanoff": true,
}

// DiscoverModules lists every .lean file under root as a module, optionally
// filtered by a regex on the module name. Results are sorted by name so the
// output does not depend on directory iteration order.
func DiscoverModules(root string, pattern string) ([]Module, error) {
	var re *regexp.Regexp
	if pattern != "" {
		var err error
		if re, err = regexp.Compile(pattern); err != nil {
			return nil, fmt.Errorf("leanoff: invalid --filter regex: %v", err)
		}
	}
	mods := map[string]string{}
	_ = filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // os.walk silently skips unreadable directories
		}
		if d.IsDir() {
			if path != root && skipDirs[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if !strings.HasSuffix(d.Name(), ".lean") {
			return nil
		}
		rel, err := filepath.Rel(root, path)
		if err != nil {
			return nil
		}
		name := moduleNameFromRel(rel)
		if re != nil && !re.MatchString(name) {
			return nil
		}
		mods[name] = path
		return nil
	})
	names := make([]string, 0, len(mods))
	for n := range mods {
		names = append(names, n)
	}
	sort.Strings(names)
	out := make([]Module, 0, len(names))
	for _, n := range names {
		out = append(out, Module{Name: n, Path: mods[n]})
	}
	return out, nil
}

// moduleNameFromRel turns "Dir/Sub/Mod.lean" into "Dir.Sub.Mod", mirroring
// the Python reference: rel.with_suffix("").replace(os.sep, "."). A leading
// dot in the final component is not a suffix, so a file literally named
// ".lean" keeps its name.
func moduleNameFromRel(rel string) string {
	return strings.ReplaceAll(stripPySuffix(rel), string(filepath.Separator), ".")
}

func stripPySuffix(p string) string {
	lastDot := strings.LastIndex(p, ".")
	lastSep := strings.LastIndexAny(p, `/\`)
	if lastDot > lastSep+1 {
		return p[:lastDot]
	}
	return p
}

// ParseImports fills mod.Deps with the module's project-internal imports:
// every `import X` line whose target is a discovered module and not the
// module itself.
func ParseImports(mod *Module, known map[string]bool) {
	data, err := os.ReadFile(mod.Path)
	if err != nil {
		return
	}
	var deps []string
	for _, line := range pySplitLines(string(data)) {
		m := importRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		if known[m[1]] && m[1] != mod.Name {
			deps = append(deps, m[1])
		}
	}
	mod.Deps = deps
}

// TopoLevels groups modules into dependency levels; each level is
// parallelizable. Returns an error on import cycles instead of exiting.
func TopoLevels(mods []Module) ([][]Module, error) {
	remaining := make(map[string]map[string]bool, len(mods))
	byName := make(map[string]Module, len(mods))
	for _, m := range mods {
		s := make(map[string]bool, len(m.Deps))
		for _, d := range m.Deps {
			s[d] = true
		}
		remaining[m.Name] = s
		byName[m.Name] = m
	}
	done := map[string]bool{}
	var levels [][]Module
	for len(remaining) > 0 {
		var ready []string
		for n, ds := range remaining {
			ok := true
			for d := range ds {
				if !done[d] {
					ok = false
					break
				}
			}
			if ok {
				ready = append(ready, n)
			}
		}
		if len(ready) == 0 {
			names := make([]string, 0, len(remaining))
			for n := range remaining {
				names = append(names, n)
			}
			sort.Strings(names)
			return nil, fmt.Errorf("leanoff: import cycle among: %s", strings.Join(names, ", "))
		}
		sort.Strings(ready)
		level := make([]Module, 0, len(ready))
		for _, n := range ready {
			level = append(level, byName[n])
			done[n] = true
			delete(remaining, n)
		}
		levels = append(levels, level)
	}
	return levels, nil
}

File: cpp/module.cpp
// module.cpp — see module.h. Mirrors internal/leanoff/module.go and the
// discover_modules / parse_imports / topo_levels functions of leanoff.py.
#include "module.h"

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <functional>
#include <map>
#include <memory>
#include <stdexcept>

#include "pycompat.h"
#include "pyre.h"

namespace leanoff {
namespace {

namespace fs = std::filesystem;

const char* const kSkipDirs[] = {
    ".lake", ".git", ".vscode", "node_modules", "dist", "out", "build", ".leanoff",
};

bool isSkipDir(const std::string& name) {
    for (const char* d : kSkipDirs) {
        if (name == d) return true;
    }
    return false;
}

std::string pathU8(const fs::path& p) {
    return p.u8string();
}

}  // namespace

std::string moduleNameFromRel(const std::string& rel) {
    // stripPySuffix: Python's with_suffix("") — drop the last dot-suffix only
    // when it lies after the last separator (a leading dot is not a suffix).
    int64_t lastDot = -1;
    int64_t lastSep = -1;
    for (size_t i = 0; i < rel.size(); i++) {
        if (rel[i] == '.') lastDot = static_cast<int64_t>(i);
        if (rel[i] == '/' || rel[i] == '\\') lastSep = static_cast<int64_t>(i);
    }
    std::string stripped = rel;
    if (lastDot > lastSep + 1) stripped = rel.substr(0, static_cast<size_t>(lastDot));
    for (char& c : stripped) {
        if (c == '/' || c == '\\') c = '.';
    }
    return stripped;
}

std::vector<Module> discoverModules(const std::string& root, const std::string& filter) {
    std::unique_ptr<Regexp> re;
    if (!filter.empty()) {
        try {
            re = Regexp::compile(filter);
        } catch (const std::exception& e) {
            throw std::runtime_error("leanoff: invalid --filter regex: " + std::string(e.what()));
        }
    }
    std::map<std::string, std::string> mods;  // name -> path, sorted by name
    const fs::path rootP = fs::u8path(root);

    // os.walk / WalkDir: children in SKIP_DIRS are pruned anywhere in the
    // tree (the root itself is never skipped); unreadable directories are
    // skipped silently. relDir tracks the path relative to root so module
    // names never need fs::relative's canonicalization.
    std::function<void(const fs::path&, const fs::path&)> walk =
        [&](const fs::path& dir, const fs::path& relDir) {
            std::error_code ec;
            fs::directory_iterator it(dir, ec);
            if (ec) return;  // unreadable directory: skip, keep walking siblings
            std::vector<fs::path> subdirs;
            for (const auto& entry : it) {
                std::error_code ec2;
                // Symlinks to directories are not descended (os.walk with
                // followlinks=False, WalkDir's non-following d.IsDir()).
                bool isDir = !entry.is_symlink(ec2) && entry.is_directory(ec2);
                std::string base = pathU8(entry.path().filename());
                if (isDir) {
                    if (isSkipDir(base)) continue;
                    subdirs.push_back(entry.path());
                    continue;
                }
                if (base.size() <= 5 ||
                    base.compare(base.size() - 5, 5, ".lean") != 0)
                    continue;
                fs::path rel = relDir.empty() ? entry.path().filename()
                                              : relDir / entry.path().filename();
                std::string name = moduleNameFromRel(pathU8(rel));
                if (re && !re->find(name).has_value()) continue;
                mods[name] = pathU8(entry.path());
            }
            for (const auto& d : subdirs) {
                fs::path subRel = relDir.empty() ? d.filename() : relDir / d.filename();
                walk(d, subRel);
            }
        };
    walk(rootP, fs::path());

    std::vector<Module> out;
    out.reserve(mods.size());
    for (const auto& kv : mods) {
        out.push_back(Module{kv.first, kv.second, {}});
    }
    return out;
}

void parseImports(Module& mod, const std::set<std::string>& known) {
    std::ifstream in(fs::u8path(mod.path), std::ios::binary);
    if (!in) return;  // unreadable file: no deps, like OSError in Python
    std::string data((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
    static const std::unique_ptr<Regexp> importRe = [] {
        return Regexp::compile("^\\s*import\\s+([A-Za-z_][A-Za-z0-9_.]*)");
    }();
    std::vector<std::string> deps;
    for (const auto& line : pySplitLines(data)) {
        auto m = importRe->matchString(line);
        if (!m) continue;
        std::string dep = m->group(1);
        if (known.count(dep) && dep != mod.name) deps.push_back(dep);
    }
    mod.deps = std::move(deps);
}

std::vector<std::vector<Module>> topoLevels(const std::vector<Module>& mods) {
    std::map<std::string, std::set<std::string>> remaining;
    std::map<std::string, Module> byName;
    for (const auto& m : mods) {
        remaining[m.name] = std::set<std::string>(m.deps.begin(), m.deps.end());
        byName[m.name] = m;
    }
    std::set<std::string> done;
    std::vector<std::vector<Module>> levels;
    while (!remaining.empty()) {
        std::vector<std::string> ready;
        for (const auto& kv : remaining) {
            bool ok = true;
            for (const auto& d : kv.second) {
                if (!done.count(d)) {
                    ok = false;
                    break;
                }
            }
            if (ok) ready.push_back(kv.first);
        }
        if (ready.empty()) {
            std::vector<std::string> names;
            for (const auto& kv : remaining) names.push_back(kv.first);
            std::sort(names.begin(), names.end());
            std::string joined;
            for (size_t i = 0; i < names.size(); i++) {
                if (i > 0) joined += ", ";
                joined += names[i];
            }
            throw std::runtime_error("leanoff: import cycle among: " + joined);
        }
        std::sort(ready.begin(), ready.end());
        std::vector<Module> level;
        for (const auto& n : ready) {
            level.push_back(byName[n]);
            done.insert(n);
            remaining.erase(n);
        }
        levels.push_back(std::move(level));
    }
    return levels;
}

}  // namespace leanoff
